lambdas/pkgpush/src/t4_lambda_pkgpush/__init__.py

```diff
@@ -228,138 +109,196 @@ def from_quilt_exception(cls, qe: quilt3.util.QuiltException):
+    )
+
+    parsed = json.load(resp["Payload"])
+    logger.info("response from hash lambda: %s", parsed)
+
+    if "FunctionError" in resp:
+        raise PkgpushException("S3HashLambdaUnhandledError", parsed)
+
+    if "error" in parsed:
+        raise PkgpushException("S3HashLambdaError", parsed["error"])
+
+    return ChecksumResult(**parsed["result"]).checksum
+
+
+def calculate_pkg_entry_hash(
+    pkg_entry: quilt3.packages.PackageEntry,
+    credentials: AWSCredentials,
+):
+    logger.info("calculate_pkg_entry_hash(%s): start", pkg_entry.physical_key)
+    pkg_entry.hash = invoke_hash_lambda(pkg_entry.physical_key, credentials).dict()
+    logger.info("calculate_pkg_entry_hash(%s): done", pkg_entry.physical_key)
+
+
+def calculate_pkg_hashes(pkg: quilt3.Package):
     entries = []
     for lk, entry in pkg.walk():
         if entry.hash is not None:
             continue
+        assert isinstance(entry.size, int)
         if entry.size > S3_HASH_LAMBDA_MAX_FILE_SIZE_BYTES:
-            raise PkgpushException("FileTooLargeForHashing", {
-                "logical_key": lk,
-                "size": entry.size,
-                "max_size": S3_HASH_LAMBDA_MAX_FILE_SIZE_BYTES,
-            })
+            raise PkgpushException(
+                "FileTooLargeForHashing",
+                {
+                    "logical_key": lk,
+                    "size": entry.size,
+                    "max_size": S3_HASH_LAMBDA_MAX_FILE_SIZE_BYTES,
+                },
+            )
 
         entries.append(entry)
 
-    user_s3 = boto_session.client("s3")
-
-    @functools.lru_cache(maxsize=None)
-    def get_region_for_bucket(bucket: str) -> str:
-        try:
-            resp = user_s3.head_bucket(Bucket=bucket)
-        except botocore.exceptions.ClientError as e:
-            resp = e.response
-            if resp["Error"]["Code"] == "404":
-                raise
-        return resp["ResponseMetadata"]["HTTPHeaders"]["x-amz-bucket-region"]
-
-    @functools.lru_cache(maxsize=None)
-    def get_s3_client_for_region(region: str):
-        return boto_session.client("s3", region_name=region, config=botocore.client.Config(signature_version="s3v4"))
-
-    def get_client_for_bucket(bucket: str):
-        return get_s3_client_for_region(get_region_for_bucket(bucket))
+    logger.info("calculate_pkg_hashes: %s entries to hash", len(entries))
 
-    with concurrent.futures.ThreadPoolExecutor(max_workers=S3_HASH_LAMBDA_CONCURRENCY) as pool:
+    with concurrent.futures.ThreadPoolExecutor(
+        max_workers=S3_HASH_LAMBDA_CONCURRENCY
+    ) as pool:
+        credentials = AWSCredentials.from_boto_session(user_boto_session.get())
         fs = [
-            pool.submit(calculate_pkg_entry_hash, get_client_for_bucket, entry)
+            pool.submit(calculate_pkg_entry_hash, entry, credentials)
             for entry in entries
         ]
         for f in concurrent.futures.as_completed(fs):
             f.result()
```

Comment:
Not quite related to this PR, but I think we can finish faster if we schedule larger tasks first i.e. if we sort `entries` by `physical_key.size`.